### ISSUE TITLE
[SCM-889] Jazz tag command creates snapshot in wrong workspace

### ISSUE THREAD

**Issue by olamy** [olamy/maven-scm#546]
**[Sérgio Ozaki](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=serozaki)** opened **[SCM-889](https://issues.apache.org/jira/browse/SCM-889?redirect=false)** and commented

Jazz provider tag command uses the jazz workspace set int the pom.xml scm tag. When in other jazz workspace, the snapshot is not created based on the current work.

---

**Affects:** 1.9.5, 1.10.0

**Remote Links:**
- [GitHub Pull Request #76
](https://github.com/apache/maven-scm/pull/76)

**Comment by olamy** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [fd177662edb0a9d712e87ae373373c135b959824](https://gitbox.apache.org/repos/asf?p=maven-scm.git;a=commit;h=fd177662edb0a9d712e87ae373373c135b959824).

**Comment by olamy** [comment]
**[Chris Graham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrisgwarp)** commented

Can I please have a bit more information about the usage and workflow?
The tag command, as documented here: http://maven.apache.org/scm/maven-scm-providers/maven-scm-provider-jazz/, details what happens.
The code, as designed, is meant to work with the release plugin - the most common use case, in which case the original code is correct, as the value needs to come from the pom file.

In all of my testing, the current code is correct.

I would like to know how you are using the tag command. Can you please elaborate?

**Comment by olamy** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[ChrisGWarp](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ChrisGWarp), please do not reopen issues which have been fixed. If you think this change is wrong, we should spawn a new ticket reverting it. Though, asking the original author is, of course, legit.

**Comment by olamy** [comment]
**[Chris Graham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrisgwarp)** commented

A) I don't think it has been fixed. I'm considering rolling it back. But I need to do a bit more work to review the context, which is why I wanted the original opener to comment on their usage patterns. As creating a tag is normally only ever via via the release plugin, I'd like to know where/how it's being used to expose this case.

B) By opening a new issue, we will loose the history and discussion on this issue where it needs to be relevant (ie here and not via a linked issue).

**Comment by olamy** [comment]
**[Chris Graham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrisgwarp)** commented

Here is the output of a release:prepare

```
[INFO] Checking in modified POMs...
[INFO] Executing: cmd.exe /X /C "scm create changeset --username Deb --password ***** "[maven-release-plugin] prepare release GPDB-1.0.56""
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm status --username Deb --password ***** --wide"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm changeset associate --username Deb --password ***** 2397 62"
[INFO] Executing: cmd.exe /X /C "scm status --username Deb --password ***** --wide"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm checkin --username Deb --password ***** C:\Workspaces\Apache\GPDB\pom.xml C:\Workspaces\Apache\GPDB\GPDBResources\pom.xml C:\Wor
kspaces\Apache\GPDB\GPDBWeb\pom.xml C:\Workspaces\Apache\GPDB\GPDBEAR\pom.xml"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm deliver --repository-uri https://rtc:9444/jazz --username Deb --password ***** --source GPDBWorkspace --target GPDBStream --over
write-uncommitted"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Tagging release with the label GPDB-1.0.56...
[INFO] Executing: cmd.exe /X /C "scm status --username Deb --password ***** --wide"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm create snapshot --repository-uri https://rtc:9444/jazz --username Deb --password ***** --name GPDB-1.0.56 --description "[maven-release-plugin] copy for tag GPDB-1.0.56" GPDBWorkspace"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm create workspace --repository-uri https://rtc:9444/jazz --username Deb --password ***** GPDB-1.0.56 --snapshot GPDB-1.0.56"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm deliver --repository-uri https://rtc:9444/jazz --username Deb --password ***** --source GPDB-1.0.56 --target GPDBStream"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm snapshot promote --repository-uri https://rtc:9444/jazz --username Deb --password ***** GPDBStream GPDB-1.0.56"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Transforming 'GPDB - J2EE Multi Module Project'...
[INFO] Transforming 'GPDB - Resources Project'...
[INFO] Transforming 'GPDB - WAR Project'...
[INFO]   Updating GPDBResources to 1.0.57-SNAPSHOT
[INFO] Transforming 'GPDB - EAR Project'...
[INFO]   Updating GPDBWeb to 1.0.57-SNAPSHOT
[INFO] Not removing release POMs
[INFO] Checking in modified POMs...
[INFO] Executing: cmd.exe /X /C "scm create changeset --username Deb --password ***** "[maven-release-plugin] prepare for next development iteration""
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm status --username Deb --password ***** --wide"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm changeset associate --username Deb --password ***** 2401 62"
[INFO] Executing: cmd.exe /X /C "scm status --username Deb --password ***** --wide"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm checkin --username Deb --password ***** C:\Workspaces\Apache\GPDB\pom.xml C:\Workspaces\Apache\GPDB\GPDBResources\pom.xml C:\Workspaces\Apache\GPDB\GPDBWeb\pom.xml C:\Workspaces\Apache\GPDB\GPDBEAR\pom.xml"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Executing: cmd.exe /X /C "scm deliver --repository-uri https://rtc:9444/jazz --username Deb --password ***** --source GPDBWorkspace --target GPDBStream --over
write-uncommitted"
[INFO] Working directory: C:\Workspaces\Apache\GPDB
[INFO] Release preparation complete.
```

and the checkout (or load in jazz SCM terms) is:

```
[INFO] --- maven-release-plugin:3.0.0-SNAPSHOT:perform (default-cli) @ GPDB ---
[INFO] Checking out the project to perform the release ...
[INFO] Executing: cmd.exe /X /C "scm load --force --repository-uri https://rtc:9444/jazz --username Deb --password ***** --dir C:\tmp\maven GPDB-1.0.56"
[INFO] Working directory: C:\tmp\maven
[INFO] Invoking perform goals in directory C:\tmp\maven\GPDB
[INFO] Executing goals 'deploy'...
```

The reason the original code used getRepositoryWorkspace() which is parsed from the pom before anything else happens, over the change to getWorkspace(), which is parsed from the output of a ```
scm status

```command, is that the status command may not have been called.

There is an (from memory, I will have to cross check all of the calls across all commands) an implicit assumption that the repository workspace matches the value in the pom.

What the reporter here appears to have done, is switched/changed workspaces - without reflecting those changes in the pom - which is there the inconsistency has come from.

The deliver command might also fail as a result of this.

In short, I don't think it's a valid error, and the user's workflow is wrong. So I need them to comment on how it is being used, and why they think the original implementation is wrong.

It is the last value to the create snapshot command in question:
```

scm create snapshot --repository-uri https://rtc:9444/jazz --username Deb --password ***** --name GPDB-1.0.56 --description "[maven-release-plugin] copy for tag GPDB-1.0.56" GPDBWorkspace

```

If that is different to the value in the pom, the source parameter to the deliver command:
```

scm deliver --repository-uri https://rtc:9444/jazz --username Deb --password ***** --source GPDBWorkspace --target GPDBStream --overwrite-uncommitted

```

also needs to be the same value as above.
```

**Comment by olamy** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[ChrisGWarp](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ChrisGWarp), I'd like to close this ticket and kindly ask you to create a new one for the regression.

**Comment by olamy** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please do not reopen closed issued when the release has been performed. Open a new issue please.

**Comment by olamy** [comment]
**[Chris Graham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrisgwarp)** commented

What this issue boils down to is this:

Should the tag implementation use what is defined in the pom, or what is currently on-disk, checkedout?

I've written it based on what is in the pom (which is how I thought that the other providers did it). This issue provides a fix for the latter.

How have the other providers handled this?

[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy), [Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte) [Hervé Boutemy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hboutemy) Thoughts on how this issue should be resolved?

**Comment by olamy** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

> Should the tag implementation use what is defined in the pom, or what is currently on-disk, checkedout?

I'd say pom by default. Not sure if this could be compared with [GIT's fetch/push](https://maven.apache.org/scm/git.html), but you could extend the scm url for jazz if that helps.

**Comment by olamy** [comment]
**[Sérgio Ozaki](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=serozaki)** commented

Hello, let me try to explain why I have opened this topic and proposed this change.

Comparing to git, the jazz workspace is equivalent to a git branch. There is not really a concept of repository in jazz like there is in git.

In my use case I was trying to use the scm plug-in with the release plug-in. I had a "master" workspace and some other previous "release" workspaces that I had to support.  The "master" workspace url was the one I kept in pom.xml and I didn't want to change it in the "release" workspaces, because it makes the merges a litlle bit easier.

So when performing a release in the "release" workspace, the change sets (commits) with the release version and the next development version are created, but the release tag is created as a snapshot of the "master". So, the perform goal of the release plug-in checks out the "master" workspace and builds it instead of the "release".

At first, the change I proposed made sense to me, but if you decide, I can revert it.

**Comment by olamy** [comment]
**[Chris Graham](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=chrisgwarp)** commented

Thank you for responding [Sérgio Ozaki](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=serozaki)!

The code was originally written to honour the pom.xml, and I think that it should continue to be, as this change will change the existing workflow/patterns already in place.

I wrote some documentation and suggestions on how to best use the release plugin with Jazz, read it [Using the Release Plugin with the Jazz SCM provider](http://maven.apache.org/scm/maven-scm-providers/maven-scm-provider-jazz/releases.html). That was 4 years ago (before I was using this in a CI env). In a CI env, Jenkins or even Jazz's own, I would rename the *BuildWorkspaces to *ReleaseWorkspaces to make it clearer. If you do have CI in place (and not Jazz's private builds), then I would have both the *BuildWorkspaces (for the usual CI SNAPSHOT builds) and the *ReleaseWorkspaces to perform your releases in.

If you have the flow targets set up correctly, then the changesets commits and tags should be pushed up to a stream. The Jazz equivalent of a tag is a Snapshot which is across all components in a repository workspace, which is why only one component in each *BuildWorkspace (should be renamed *ReleaseWorkspace).

The concern here about your change is that it relies upon local setup not captured in the pom, which may lead to non repeatable builds.

In your case, it's clear: the workspace in the pom does not need to be changed. It simply needs to be the release workspace, not your 'master' which should probably be a Stream anyway.

I would also caution against direct comparisons with Git, Perforce, ClearCase or other similar SCM's. Each are different, some radically so. I would encourage you to think in Jazz terms, not compare against other SCM's.
